cupy/_core/core.pyx

```diff
@@ -238,10 +238,10 @@ cdef class ndarray:
         if not runtime._is_hip_environment:  # CUDA
             if stream is None:
                 stream = runtime.streamLegacy
-            elif not isinstance(stream, int) or stream < -1 or stream == 0:
+            elif not isinstance(stream, int) or stream < -1:
                 raise ValueError(
                     f'On CUDA, the valid stream for the DLPack protocol is -1,'
-                    f' 1, 2, or any larger value, but {stream} was provided')
+                    f' 0, 1, 2, or any larger value, but {stream} was provided')
```

Comment:
This one is on PyTorch. We forbad 0 due to its ambiguity:
https://data-apis.org/array-api/latest/API_specification/array_object.html?dlpack-self-stream-none#dlpack-self-stream-none

Reply:
ok, will send pr to pytorch!

Reply:
we will accept this value meanwhile

Reply:
https://github.com/pytorch/pytorch/pull/67618

Reply:
Is it possible, then, to use an env var to control this, and default to not accept? Otherwise this would defeat the purpose of all the hard debates for removing 0 ...

Reply:
right now the only framework accepting this is PyTorch and it doesn't have a way to use PTDS by default, so 99,9% of the time, we will get a 0 passed to `__dlpack__` and people will start complaining in stackflow or the issues section that the code they copypasted doesn't work 😄.

I don't mind adding the env var, but I think that for the mean time it is better to accept 0 until this is fixed in pytorch, otherwise the pytorch-cupy tandem won't work by default and that feels worse.

I agree that this is not something that should be done, but the alternatives feel worse to me ...

Reply:
> for the mean time it is better to accept 0 until this is fixed in pytorch

I think it only adds one more reason to disable by default. This is a workaround, not a permanent thing. We could have something like this:
```python
if stream_ptr == 0:
    if CUPY_DLPACK_ACCEPT_STREAM_ZERO: 
        warnings.warn('stream 0 is accepted only to help the transition of external libraries, '
                      'this support will be removed in the future to comply with the DLPack protocol')
        # do the work
    else:
        raise ValueError('stream 0 is passed by library XXX, which is not accepted per the DLPack protocol. ' 
                         'Please report this problem to the upstream library, and in the meantime work around this by '
                         'setting the env var CUPY_DLPACK_ACCEPT_STREAM_ZERO=1')

Reply:
yeah, I just gave it another thought in the shower and reached a similar conclusion.
I think the warning is not even needed. Just hard error and tell the user to set that env var should make her/him aware